File: computationGeomety/donut.cpp
#include <iostream>
#include <math.h>
#include <vector>
using namespace std;

#define PI 13.14159
#define SCREEN_WIDTH 70
#define SCREEN_HEIGHT 70


float R1 = 1;
float R2 = 2;
float K2 = 100;
int field_of_view = 130;

float K1 = SCREEN_WIDTH * K2 * 3/(8*(R1+R2));
//float K1 = (SCREEN_WIDTH/2)/tan((field_of_view/2.0) * PI/180);


void render(float A,float B){
	vector<vector<char>> screen(SCREEN_HEIGHT,vector<char>(SCREEN_WIDTH,' '));
	float cosA = cos(A);
	float sinA = sin(A);
	float cosB = cos(B);
	float sinB = sin(B);
	float theta = 0;
	while (theta < 2 * PI){
		theta += 0.1;
		float costheta = cos(theta);
		float sintheta = sin(theta);
		
		float phi = 0;
		while(phi < 2*PI){
			phi += 0.1;
			float cosPhi = cos(phi);
			float sinPhi = sin(phi);
			
			float circleX = R2 + R1*costheta;
			float circleY = R1*sintheta;
		
			
			float x = cosPhi*circleX;
			float y = circleY;
			float z = -sinPhi*circleX;;
		    //rotate at x direction	
		   
		    x = x;
		    y = cosA * y + sin(A) *z;
		    z = -sinA * y + cosA * z + K2;
		
			int px = int(x * K1 / z) + SCREEN_WIDTH/2;
		    int py = int(y * K1/ z ) + SCREEN_HEIGHT/2;
		    
			screen[py][px] = '.';
		}
		
	}

	cout<<"\x1b[H";
	for(auto col : screen){
		for(auto pixel : col)
			cout<<pixel<<" ";
		cout<<endl;
	}
	
}



int main(){

	float A = 0;
	while(true){
		render(A,0);
		A += 0.1;
	}
	return 0;
	
}

File: computationGeomety/donut.py
import math



SCREEN_HEIGHT = 60
SCREEN_WIDTH = 60
R1 = 1
R2 = 2
K2 = 600


K1 = SCREEN_WIDTH * K2 * 3/(8*(R1+R2))

def donut(A):
	screen = []
	for i in range(SCREEN_HEIGHT):
		screen.append([" "]*SCREEN_WIDTH)
	cosA = math.cos(A)
	sinA = math.sin(A)
	
	theta = 0
	
	while theta < 2 * math.pi:
		theta += 0.1
		costheta = math.cos(theta)
		sintheta = math.sin(theta)
	
		phi = 0
		while phi < math.pi *2:
			phi += 0.01
			circleX = R2 + R1*costheta
			circleY = R1*sintheta
			circleZ = 0
			cosphi = math.cos(phi)
			sinphi = math.sin(phi)
			
			x = circleX * cosphi
			y = circleY
			z = (circleX * sinphi)
			
			x = x
			y = y * cosA + z * (-sinA)
			z = y * sinA + z * cosA + K2
			
			ooz = 1/z
			x = int(SCREEN_WIDTH/2 + K1*ooz*x)
			y = int(SCREEN_WIDTH/2 + K1*ooz*y)
			
			screen[x][y] = "."
	
	print('\x1b[H')	
	for i in range(SCREEN_HEIGHT):
		for j in range(SCREEN_WIDTH):
			print(screen[i][j], end=" ")
			
		print()
		
		
print("start")
a = 0
while True:
	donut(a)
	a += 0.2
print("end")

		
	
